ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -434,8 +437,18 @@ private boolean removeFiles(String location, ValidWriteIdList writeIdList, Compa
     return success;
   }
 
-  private boolean hasDataBelowWatermark(FileSystem fs, Path path, long highWatermark) throws IOException {
-    FileStatus[] children = fs.listStatus(path);
+  private boolean hasDataBelowWatermark(AcidDirectory acidDir, FileSystem fs, Path path, long highWatermark)
+      throws IOException {
+    Set<Path> acidPaths = new HashSet<>();
+    for (ParsedDelta delta : acidDir.getCurrentDirectories()) {
+      acidPaths.add(delta.getPath());
+    }
+    if (acidDir.getBaseDirectory() != null) {
+      acidPaths.add(acidDir.getBaseDirectory());
+    }
+    FileStatus[] children = fs.listStatus(path, p -> {
+      return !acidPaths.contains(p);
+    });
     for (FileStatus child : children) {
       if (isFileBelowWatermark(child, highWatermark)) {
```

Comment:
Commenting on the contents of isFileBelowWatermark since I can't comment there...

1. `if (!child.isDirectory()) {
      return false;
    }`
There could be original files in the table directory that should be deleted.

2. `return b.getWriteId() < highWatermark;`
the highWatermark is inclusive, so if for some reason the table directory contains:
delta_5_5
delta_1_5_v100 (minor compacted) -- this includes the data in delta_5_5.
then isFileBelowWatermark would return false but it should return true.

Reply:
1. I believe that in case there are files in the dir they already should be in the `obsolete` list; I just wanted to be conservative in this method - but I think returning true there would be correct as well
2. the `highWatermark` is inclusive; but this method's name is isBelowWatermark - so it only looks for files which are below the watermark
  w.r.t to `delta_1_5` ; I think its not below `5` because it contains data from `writeId` 5.

Reply:
1.
> I believe that in case there are files in the dir they already should be in the obsolete list

Not necessarily, because the AcidDirectory the Cleaner uses is computed based on an older txnId (cleanerWaterMark), so there is a chance its obsolete list does not contain files that should be cleaned up eventually, which is what this method is supposed to figure out. (Right?)

@deniskuzZ please correct me if I'm wrong about this since I know there have been recent changes to this logic

2. I meant that if the table dir contains:

- delta_5_5
- delta_1_5_v100 (minor compacted)

Then the cleaner should eventually remove delta_5_5, so there will be files to remove later, when the cleanerWaterMark is high enough

Reply:
after some thinking I convinced myself that you are right  :)

* I've changed to return `true` for non-directories
** in the background these should appear as `obsolete` files anyway; so it should not cause any real trouble in the scope of HIVE-25883; but it makes the method live up to its name...
* since the latest patch we are checking and excluding all the dirs the actual acid dir is using - so if we have anything below or even at the writeid level that should be considered invalid; the `nothingToCleanAfterAbortsDelta` testcase is a "complicated" case but the default case is similar to this with the new checks.

pushed a new commit to update these

Reply:
@klcopp the current logic is that if there is an open txn started before Compaction txn committed, we won't even pick this request for cleanup.

Reply:
@kgyrtkirk, could you please try the following test case (simulate legacy behavior when CQ_NEXT_TXN_ID=NULL by commenting out updateWSCommitIdAndCleanUpMetadata in CompactionTxnHandler):
````
    String dbName = "default";
    String tblName = "trfcp";
    String partName = "ds=today";
    Table t = newTable(dbName, tblName, true);
    Partition p = newPartition(t, "today");

    // block cleaner with an open txn
    long blockingTxn = openTxn();

    // minor compaction
    addBaseFile(t, p, 20L, 20);
    addDeltaFile(t, p, 21L, 21L, 1);
    addDeltaFile(t, p, 22L, 22L, 1);
    burnThroughTransactions(dbName, tblName, 22);
    CompactionRequest rqst = new CompactionRequest(dbName, tblName, CompactionType.MINOR);
    rqst.setPartitionname(partName);
    compactInTxn(rqst);
    addDeltaFile(t, p, 21, 22, 2);
    startCleaner();

    // make sure cleaner didn't remove anything, and cleaning is still queued
    List<Path> paths = getDirectories(conf, t, p);
    Assert.assertEquals("Expected 4 files after minor compaction, instead these files were present " + paths,
      4, paths.size());
    ShowCompactResponse rsp = txnHandler.showCompact(new ShowCompactRequest());
    Assert.assertEquals("Expected 1 compaction in queue, got: " + rsp.getCompacts(), 1, rsp.getCompactsSize());
    Assert.assertEquals(TxnStore.CLEANING_RESPONSE, rsp.getCompacts().get(0).getState());
    Assert.assertEquals(CompactionType.MINOR, rsp.getCompacts().get(0).getType());
````

Reply:
note, there is shitty/unreliable logic for openTxn. When running the above test, put a breakpoint in Cleaner at  cleanerWaterMark and set minTxnIdSeenOpen to 23 (it might be 1, due to artificial commit txn in TXNS table)

Reply:
+ minor change in compacted delta file that lacks visibility id
````
case DELTA: filename = AcidUtils.addVisibilitySuffix(makeDeltaDirName(minTxn, maxTxn),visibilityId); break;
````